# Detailed Team Information

From authenticated JSON
JSON mixes race-specific and season-specific data

In [19]:
import json

with open('../../data/raw/detailed_teams/4e9c9222-1995-11f1-baf9-9b6c8660a10d-0-197320986_team_1_race_1.json', 'r') as f:
    data = json.load(f)

type(data)

dict

In [20]:
data.keys()

dict_keys(['Data', 'Meta'])

In [21]:
data["Data"]

{'Value': {'mdid': 1,
  'userTeam': [{'gdrank': 661806,
    'ovrank': 183102,
    'teamno': 1,
    'teambal': 0.2,
    'teamval': 99.8,
    'gdpoints': 179.0,
    'matchday': 1,
    'ovpoints': 3343.0,
    'playerid': [{'id': '11032',
      'isfinal': 0,
      'iscaptain': 0,
      'ismgcaptain': 0,
      'playerpostion': 1},
     {'id': '111',
      'isfinal': 0,
      'iscaptain': 0,
      'ismgcaptain': 0,
      'playerpostion': 3},
     {'id': '11051',
      'isfinal': 0,
      'iscaptain': 0,
      'ismgcaptain': 0,
      'playerpostion': 4},
     {'id': '210',
      'isfinal': 0,
      'iscaptain': 0,
      'ismgcaptain': 0,
      'playerpostion': 6},
     {'id': '115',
      'isfinal': 0,
      'iscaptain': 1,
      'ismgcaptain': 0,
      'playerpostion': 2},
     {'id': '11031',
      'isfinal': 0,
      'iscaptain': 0,
      'ismgcaptain': 0,
      'playerpostion': 5},
     {'id': '28',
      'isfinal': 0,
      'iscaptain': 0,
      'ismgcaptain': 0,
      'playerpostion': 7

In [22]:
data["Data"].keys()

dict_keys(['Value', 'FeedTime'])

In [23]:
data["Data"]["Value"].keys()

dict_keys(['mdid', 'userTeam', 'retval'])

In [24]:
type(data["Data"]["Value"]["mdid"])

int

mdid is probably the race number id

In [25]:
data["Data"]["Value"]["mdid"]

1

In [26]:
type(data["Data"]["Value"]["userTeam"])

list

In [27]:
type(data["Data"]["Value"]["retval"])

int

In [28]:
data["Data"]["Value"]["retval"]

1

Understanding what is inside the list in userTeam

In [29]:
len(data["Data"]["Value"]["userTeam"])

1

In [30]:
type(data["Data"]["Value"]["userTeam"][0])

dict

In [31]:
data["Data"]["Value"]["userTeam"][0].keys()

dict_keys(['gdrank', 'ovrank', 'teamno', 'teambal', 'teamval', 'gdpoints', 'matchday', 'ovpoints', 'playerid', 'socialId', 'teamname', 'usersubs', 'boosterid', 'team_info', 'fttourgdid', 'fttourmdid', 'iswildcard', 'maxteambal', 'capplayerid', 'subsallowed', 'isaccounting', 'usersubsleft', 'extrasubscost', 'islateonboard', 'mgcapplayerid', 'race_category', 'finalfxracecat', 'finalfxraceday', 'isboostertaken', 'extradrstakengd', 'finalfixtakengd', 'isextradrstaken', 'isfinalfixtaken', 'iswildcardtaken', 'wildcardtakengd', 'autopilottakengd', 'isautopilottaken', 'islimitlesstaken', 'limitlesstakengd', 'isnonigativetaken', 'nonigativetakengd', 'finalfxnewplayerid', 'finalfxoldplayerid', 'player_swap_details', 'is_wildcard_taken_gd_id', 'inactive_driver_penality_points'])

In [32]:
team = data["Data"]["Value"]["userTeam"][0]

In [33]:
import pandas as pd

In [34]:
pd.Series(team)

gdrank                                                                        661806
ovrank                                                                        183102
teamno                                                                             1
teambal                                                                          0.2
teamval                                                                         99.8
gdpoints                                                                       179.0
matchday                                                                           1
ovpoints                                                                      3343.0
playerid                           [{'id': '11032', 'isfinal': 0, 'iscaptain': 0,...
socialId                                                                   197320986
teamname                                                            English%20Racing
usersubs                                                         

<img src="../../docs/screenshots/Misc/English_Racing_Overall_Ranking_Detailed_team_Notebook1.png"
     alt="English Racing – Overall Season Points"
     width="800">

Screenshot above shows current points and ranking on overall leaderboard for the league.

Confirms: 
**ovpoints** is the current cumulative points this season for the team

<img src="../../docs/screenshots/Misc/English_Racing_Race1_Results_Detailed_team_Notebook2.png"
     alt="English Racing – Race 1"
     width="500">

Screenshot above, showing Race 1 (Australia) results for English Racing confirms:

**teambal** (0.2) is unused money towards the cost cap

**teamval** (99.8) is the total current value of the selected assets in the team 

**maxteambal** (100) is the total possible value of the team which the player could spend. 
In this case it is Race 1 and all teams begin the year with 100k to spend on assets. 


**gdpoints** (179) is the points earned by this team in this race.

In [35]:
for key, value in team.items():
    print(key, type(value).__name__)

gdrank int
ovrank int
teamno int
teambal float
teamval float
gdpoints float
matchday int
ovpoints float
playerid list
socialId int
teamname str
usersubs int
boosterid NoneType
team_info dict
fttourgdid int
fttourmdid int
iswildcard NoneType
maxteambal float
capplayerid str
subsallowed int
isaccounting int
usersubsleft int
extrasubscost int
islateonboard int
mgcapplayerid NoneType
race_category str
finalfxracecat NoneType
finalfxraceday NoneType
isboostertaken int
extradrstakengd NoneType
finalfixtakengd NoneType
isextradrstaken int
isfinalfixtaken int
iswildcardtaken int
wildcardtakengd int
autopilottakengd NoneType
isautopilottaken int
islimitlesstaken int
limitlesstakengd NoneType
isnonigativetaken int
nonigativetakengd int
finalfxnewplayerid NoneType
finalfxoldplayerid NoneType
player_swap_details NoneType
is_wildcard_taken_gd_id NoneType
inactive_driver_penality_points int


Need to further examine team_info and player_id which seem to be further nested dictionaries.

In [36]:
team["team_info"]

{'teamBal': 0.2,
 'teamVal': 99.8,
 'maxTeambal': 100.0,
 'subsallowed': 0,
 'userSubsleft': 0}

In [37]:
team["team_info"].keys()

dict_keys(['teamBal', 'teamVal', 'maxTeambal', 'subsallowed', 'userSubsleft'])

team_info seems to contain the same information as the parent level.

In [38]:
team["playerid"]

[{'id': '11032',
  'isfinal': 0,
  'iscaptain': 0,
  'ismgcaptain': 0,
  'playerpostion': 1},
 {'id': '111',
  'isfinal': 0,
  'iscaptain': 0,
  'ismgcaptain': 0,
  'playerpostion': 3},
 {'id': '11051',
  'isfinal': 0,
  'iscaptain': 0,
  'ismgcaptain': 0,
  'playerpostion': 4},
 {'id': '210',
  'isfinal': 0,
  'iscaptain': 0,
  'ismgcaptain': 0,
  'playerpostion': 6},
 {'id': '115',
  'isfinal': 0,
  'iscaptain': 1,
  'ismgcaptain': 0,
  'playerpostion': 2},
 {'id': '11031',
  'isfinal': 0,
  'iscaptain': 0,
  'ismgcaptain': 0,
  'playerpostion': 5},
 {'id': '28',
  'isfinal': 0,
  'iscaptain': 0,
  'ismgcaptain': 0,
  'playerpostion': 7}]

In [39]:
assets = pd.DataFrame(team["playerid"])

assets

,id,isfinal,iscaptain,ismgcaptain,playerpostion
0,11032,0,0,0,1
1,111,0,0,0,3
2,11051,0,0,0,4
3,210,0,0,0,6
4,115,0,1,0,2
5,11031,0,0,0,5
6,28,0,0,0,7


playerid indicates the assets (drivers and constructors) selected by this league member for this race.

**iscaptain** indicates the asset chosen for x2 point boost this race.

**playerpostion** [sic] is unclear. In the screenshot for this race 11032 (1) (Hadjar) appears second on the list, and 115 (2) (Leclerc) appears top of the list so there is a discrepancy here.

## Expanding to include more team x race combinations

In [40]:
import os

In [41]:
raw_dir = "../../data/raw/detailed_teams"

In [42]:
os.listdir(raw_dir)[:5]

['00190f2c-133c-11f1-9186-ede0d682ac23-0-231965328_team_1_race_6.json',
 'f88ea14a-1886-11f1-bbd2-1f9903bd5a74-0-218960172_team_1_race_4.json',
 'c3e1deaa-196f-11f1-a45e-1704308c412d-0-232549060_team_1_race_3.json',
 'f88ea14a-1886-11f1-bbd2-1f9903bd5a74-0-218960172_team_1_race_10.json',
 '354057d8-142b-11f1-9a99-a304052de58e-0-232004086_team_1_race_7.json']

In [43]:
len(os.listdir(raw_dir))

225

In [63]:
records = []

for filename in os.listdir(raw_dir):
    file_path = os.path.join(raw_dir, filename)

    with open(file_path, "r") as f:
        data = json.load(f)
    
    team = data["Data"]["Value"]["userTeam"][0]
    records.append(team)

In [64]:
len(records)

225

In [65]:
type(records)

list

In [66]:
type(records[0])

dict

Creating a DataFrame with each row representing a league member's team for a specific race.

In [67]:
teams_df = pd.DataFrame(records)

In [68]:
teams_df.shape

(225, 46)

In [69]:
teams_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 225 entries, 0 to 224
Data columns (total 46 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   gdrank                           225 non-null    int64  
 1   ovrank                           225 non-null    int64  
 2   teamno                           225 non-null    int64  
 3   teambal                          225 non-null    float64
 4   teamval                          128 non-null    float64
 5   gdpoints                         225 non-null    float64
 6   matchday                         225 non-null    int64  
 7   ovpoints                         225 non-null    float64
 8   playerid                         225 non-null    object 
 9   socialId                         225 non-null    int64  
 10  teamname                         225 non-null    object 
 11  usersubs                         225 non-null    int64  
 12  boosterid             

In [70]:
teams_df.head()

,gdrank,ovrank,teamno,teambal,teamval,gdpoints,matchday,ovpoints,playerid,socialId,...,isautopilottaken,islimitlesstaken,limitlesstakengd,isnonigativetaken,nonigativetakengd,finalfxnewplayerid,finalfxoldplayerid,player_swap_details,is_wildcard_taken_gd_id,inactive_driver_penality_points
0,506971,30850,1,1.0,110.5,144.0,1,3884.0,"[{'id': '110', 'isfinal': 0, 'iscaptain': 0, '...",231965328,...,0,1,6.0,0,NaN,None,None,None,NaN,0
1,947170,914140,1,0.5,NaN,179.0,1,2659.0,"[{'id': '11059', 'isfinal': 0, 'iscaptain': 0,...",218960172,...,0,0,NaN,0,NaN,None,None,None,NaN,0
2,271003,193918,1,3.6,95.4,207.0,1,3324.0,"[{'id': '11031', 'isfinal': 0, 'iscaptain': 0,...",232549060,...,4,0,NaN,0,NaN,None,None,None,NaN,0
3,1338589,914140,1,0.5,NaN,148.0,1,2659.0,"[{'id': '11059', 'isfinal': 0, 'iscaptain': 0,...",218960172,...,0,0,NaN,0,NaN,None,None,None,NaN,0
4,1343279,43633,1,1.0,107.5,99.0,1,3799.0,"[{'id': '11051', 'isfinal': 0, 'iscaptain': 0,...",232004086,...,0,1,6.0,0,NaN,None,None,None,NaN,0


Understanding why we have a duplicate socialid for the same race:

In [71]:
teams_df[
    teams_df["socialId"] == 218960172
][
    ["socialId", "teamno", "teamname", "matchday", "gdpoints"]
].sort_values(["teamno", "matchday"])

,socialId,teamno,teamname,matchday,gdpoints
1,218960172,1,BoxBoxBrigade,1,179.0
3,218960172,1,BoxBoxBrigade,1,148.0
19,218960172,1,BoxBoxBrigade,1,120.0
32,218960172,1,BoxBoxBrigade,1,304.0
49,218960172,1,BoxBoxBrigade,1,158.0
50,218960172,1,BoxBoxBrigade,1,161.0
65,218960172,1,BoxBoxBrigade,1,376.0
95,218960172,1,BoxBoxBrigade,1,178.0
133,218960172,1,BoxBoxBrigade,1,91.0
139,218960172,1,BoxBoxBrigade,1,199.0


15 entries seem to correspond to the 15 races so far this year, even though the matchday is the same for each entry. We had believed matchday would correspond to the race number, but this does not seem to be the case.

In [72]:
data["Data"]["Value"]["mdid"]

3

In [73]:
with open("../../data/raw/detailed_teams/f88ea14a-1886-11f1-bbd2-1f9903bd5a74-0-218960172_team_1_race_10.json", "r") as f:
    test_data = json.load(f)

In [74]:
test_data["Data"]["Value"]["mdid"]

10

In [75]:
test_data["Data"]["Value"]["userTeam"][0]["matchday"]

1

So we need to keep **mdid** from the parent level to distinguish between the different races for the same team.

In [76]:
records = []

for filename in os.listdir(raw_dir):
    file_path = os.path.join(raw_dir, filename)

    with open(file_path, "r") as f:
        data = json.load(f)

    team = data["Data"]["Value"]["userTeam"][0].copy()
    team["race_number"] = data["Data"]["Value"]["mdid"]

    records.append(team)

In [77]:
teams_df = pd.DataFrame(records)

In [78]:
teams_df.shape

(225, 47)

In [79]:
teams_df["race_number"].nunique()

15

In [80]:
teams_df[
    teams_df["socialId"] == 218960172
][
    ["socialId", "teamno", "teamname", "race_number",
     "matchday", "gdpoints"]
].sort_values("race_number")

,socialId,teamno,teamname,race_number,matchday,gdpoints
139,218960172,1,BoxBoxBrigade,1,1,199.0
65,218960172,1,BoxBoxBrigade,2,1,376.0
95,218960172,1,BoxBoxBrigade,3,1,178.0
1,218960172,1,BoxBoxBrigade,4,1,179.0
50,218960172,1,BoxBoxBrigade,5,1,161.0
180,218960172,1,BoxBoxBrigade,6,1,6.0
211,218960172,1,BoxBoxBrigade,7,1,104.0
19,218960172,1,BoxBoxBrigade,8,1,120.0
32,218960172,1,BoxBoxBrigade,9,1,304.0
3,218960172,1,BoxBoxBrigade,10,1,148.0


## Continuing understanding which columns are race-specific vs cumulative season 

In [81]:
teams_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 225 entries, 0 to 224
Data columns (total 47 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   gdrank                           225 non-null    int64  
 1   ovrank                           225 non-null    int64  
 2   teamno                           225 non-null    int64  
 3   teambal                          225 non-null    float64
 4   teamval                          128 non-null    float64
 5   gdpoints                         225 non-null    float64
 6   matchday                         225 non-null    int64  
 7   ovpoints                         225 non-null    float64
 8   playerid                         225 non-null    object 
 9   socialId                         225 non-null    int64  
 10  teamname                         225 non-null    object 
 11  usersubs                         225 non-null    int64  
 12  boosterid             

In [ ]:
# Remove the nested columns

scalar_df = teams_df.drop(
    columns=["playerid", "team_info", "player_swap_details"]
)

In [83]:
scalar_df.nunique().sort_values()

islateonboard                        1
extrasubscost                        1
fttourmdid                           1
fttourgdid                           1
race_category                        1
finalfxracecat                       1
finalfxraceday                       1
wildcardtakengd                      1
matchday                             1
finalfxnewplayerid                   1
finalfxoldplayerid                   1
finalfixtakengd                      1
isaccounting                         2
isextradrstaken                      2
isfinalfixtaken                      2
iswildcardtaken                      2
isboostertaken                       2
autopilottakengd                     2
isautopilottaken                     2
iswildcard                           2
islimitlesstaken                     2
isnonigativetaken                    2
teamno                               2
is_wildcard_taken_gd_id              2
inactive_driver_penality_points      3
subsallowed              

In [86]:
# Examining one team across 15 races to determine static values

ramenracing = teams_df[
    teams_df["teamname"] == "Ramen%20Face%20Racing"
].sort_values("race_number")

In [87]:
ramenracing[
    [
        "race_number",
        "gdpoints",
        "gdrank",
        "teamval",
        "teambal",
        "ovpoints",
        "ovrank",
        "usersubs",
        "usersubsleft"
    ]
]

,race_number,gdpoints,gdrank,teamval,teambal,ovpoints,ovrank,usersubs,usersubsleft
31,1,259.0,37437,99.8,0.2,4307.0,1824,0,0
194,2,481.0,26748,103.1,0.2,4307.0,1824,0,2
200,3,178.0,646682,105.3,0.1,4307.0,1824,2,0
115,4,193.0,718941,107.4,1.3,4307.0,1824,2,0
159,5,358.0,3149,109.7,0.7,4307.0,1824,2,0
54,6,214.0,109933,110.5,0.0,4307.0,1824,1,1
108,7,107.0,1168292,109.7,1.7,4307.0,1824,3,0
129,8,232.0,48502,111.6,1.9,4307.0,1824,1,1
146,9,286.0,441456,113.6,2.0,4307.0,1824,1,2
171,10,252.0,12162,117.3,0.6,4307.0,1824,2,1


In [88]:
# Compare to a team that has not been changed since early in the season

boxbox = teams_df[
    teams_df["socialId"] == 218960172
].sort_values("race_number")

In [89]:
boxbox[
    [
        "race_number",
        "gdpoints",
        "gdrank",
        "teamval",
        "teambal",
        "ovpoints",
        "ovrank",
        "usersubs",
        "usersubsleft"
    ]
]

,race_number,gdpoints,gdrank,teamval,teambal,ovpoints,ovrank,usersubs,usersubsleft
139,1,199.0,457524,99.9,0.1,2659.0,914140,0,0
65,2,376.0,246824,100.8,0.5,2659.0,914140,1,1
95,3,178.0,646682,NaN,0.5,2659.0,914140,0,0
1,4,179.0,947170,NaN,0.5,2659.0,914140,0,0
50,5,161.0,1009017,NaN,0.5,2659.0,914140,0,0
180,6,6.0,1880037,NaN,0.5,2659.0,914140,0,0
211,7,104.0,1235626,NaN,0.5,2659.0,914140,0,0
19,8,120.0,1570866,NaN,0.5,2659.0,914140,0,0
32,9,304.0,244228,NaN,0.5,2659.0,914140,0,0
3,10,148.0,1338589,NaN,0.5,2659.0,914140,0,0


**teamval** does not seem to update if no changes/transfers were made to the team the previous week. This is a little surprising seeing as the assets in the team continue to change from race-to-race so what we see here may not be an accurate reflection of the current value of the team's assets.

## Understanding Chip Columns

In [90]:
chip_columns = [
    "race_number",
    "boosterid",
    "isboostertaken",
    "isextradrstaken",
    "extradrstakengd",
    "isfinalfixtaken",
    "finalfixtakengd",
    "iswildcardtaken",
    "wildcardtakengd",
    "isautopilottaken",
    "autopilottakengd",
    "islimitlesstaken",
    "limitlesstakengd",
    "isnonigativetaken",
    "nonigativetakengd",
    "is_wildcard_taken_gd_id"
]

In [92]:
# this player has not used any chips this season
boxbox[chip_columns]

,race_number,boosterid,isboostertaken,isextradrstaken,extradrstakengd,isfinalfixtaken,finalfixtakengd,iswildcardtaken,wildcardtakengd,isautopilottaken,autopilottakengd,islimitlesstaken,limitlesstakengd,isnonigativetaken,nonigativetakengd,is_wildcard_taken_gd_id
139,1,NaN,0.0,0,NaN,0,NaN,0,0,0,NaN,0,NaN,0,NaN,NaN
65,2,NaN,0.0,0,NaN,0,NaN,0,0,0,NaN,0,NaN,0,NaN,NaN
95,3,NaN,NaN,0,NaN,0,NaN,0,0,0,NaN,0,NaN,0,NaN,NaN
1,4,NaN,NaN,0,NaN,0,NaN,0,0,0,NaN,0,NaN,0,NaN,NaN
50,5,NaN,NaN,0,NaN,0,NaN,0,0,0,NaN,0,NaN,0,NaN,NaN
180,6,NaN,NaN,0,NaN,0,NaN,0,0,0,NaN,0,NaN,0,NaN,NaN
211,7,NaN,NaN,0,NaN,0,NaN,0,0,0,NaN,0,NaN,0,NaN,NaN
19,8,NaN,NaN,0,NaN,0,NaN,0,0,0,NaN,0,NaN,0,NaN,NaN
32,9,NaN,NaN,0,NaN,0,NaN,0,0,0,NaN,0,NaN,0,NaN,NaN
3,10,NaN,NaN,0,NaN,0,NaN,0,0,0,NaN,0,NaN,0,NaN,NaN


In [93]:
# this player has used a number of chips
ramenracing[chip_columns]

,race_number,boosterid,isboostertaken,isextradrstaken,extradrstakengd,isfinalfixtaken,finalfixtakengd,iswildcardtaken,wildcardtakengd,isautopilottaken,autopilottakengd,islimitlesstaken,limitlesstakengd,isnonigativetaken,nonigativetakengd,is_wildcard_taken_gd_id
31,1,NaN,0.0,6,12.0,0,NaN,0,0,0,NaN,1,2.0,5,5.0,NaN
194,2,1.0,1.0,6,12.0,0,NaN,0,0,0,NaN,1,2.0,5,5.0,NaN
200,3,NaN,0.0,6,12.0,0,NaN,0,0,0,NaN,1,2.0,5,5.0,NaN
115,4,NaN,0.0,6,12.0,0,NaN,0,0,0,NaN,1,2.0,5,5.0,NaN
159,5,5.0,1.0,6,12.0,0,NaN,0,0,0,NaN,1,2.0,5,5.0,NaN
54,6,NaN,0.0,6,12.0,0,NaN,0,0,0,NaN,1,2.0,5,5.0,NaN
108,7,NaN,0.0,6,12.0,0,NaN,0,0,0,NaN,1,2.0,5,5.0,NaN
129,8,NaN,0.0,6,12.0,0,NaN,0,0,0,NaN,1,2.0,5,5.0,NaN
146,9,NaN,0.0,6,12.0,0,NaN,0,0,0,NaN,1,2.0,5,5.0,NaN
171,10,NaN,0.0,6,12.0,0,NaN,0,0,0,NaN,1,2.0,5,5.0,NaN


Screenshot below shows Ramen Racing team for Race 12 (Netherlands) where x3 boost (extrasdrs) was used. 

This also confirms:
- `usersubs = 4` and `usersubsleft = -1` coincide with the interface showing one additional transfer and a -10 point penalty.


<img src="../../docs/screenshots/Misc/Ramen_Racing_Race12_x3Boost_Notebook3.png"
     alt="Ramen Face Racing – Race 12 x3 Boost"
     width="500">

Looking more closely at the teaminfo for Ramen Racing for Race 12.

In [96]:
ramen_race12 = ramenracing[
    ramenracing["race_number"] == 12
].iloc[0]

ramen_race12["playerid"]

[{'id': '11031',
  'isfinal': 0,
  'iscaptain': 0,
  'ismgcaptain': 0,
  'playerpostion': 4},
 {'id': '111',
  'isfinal': 0,
  'iscaptain': 0,
  'ismgcaptain': 0,
  'playerpostion': 3},
 {'id': '11161',
  'isfinal': 0,
  'iscaptain': 1,
  'ismgcaptain': 0,
  'playerpostion': 1},
 {'id': '117',
  'isfinal': 0,
  'iscaptain': 0,
  'ismgcaptain': 1,
  'playerpostion': 2},
 {'id': '129',
  'isfinal': 0,
  'iscaptain': 0,
  'ismgcaptain': 0,
  'playerpostion': 5},
 {'id': '25',
  'isfinal': 0,
  'iscaptain': 0,
  'ismgcaptain': 0,
  'playerpostion': 7},
 {'id': '27',
  'isfinal': 0,
  'iscaptain': 0,
  'ismgcaptain': 0,
  'playerpostion': 6}]

In [98]:
pd.DataFrame(ramen_race12["playerid"])

,id,isfinal,iscaptain,ismgcaptain,playerpostion
0,11031,0,0,0,4
1,111,0,0,0,3
2,11161,0,1,0,1
3,117,0,0,1,2
4,129,0,0,0,5
5,25,0,0,0,7
6,27,0,0,0,6


In [99]:
ramenracing[
    ramenracing["race_number"] == 12
][
    [
        "race_number",
        "boosterid",
        "capplayerid",
        "mgcapplayerid",
        "isboostertaken",
        "isextradrstaken",
        "extradrstakengd"
    ]
]

,race_number,boosterid,capplayerid,mgcapplayerid,isboostertaken,isextradrstaken,extradrstakengd
15,12,6.0,11161,117,1.0,6,12.0


In [105]:
# testing the hypothesis that isdrstaken = 6 corresponds to the playerpostion in teaminfo, which in this case is 27. if correct 27 should be Norris
# cross reference the asset id to see which driver it is.
assets_race12 = pd.read_csv(
    "../../data/processed/asset_snapshot_race_12.csv"
)

In [104]:
assets_race12[
    assets_race12["asset_id"] == 27
]

,season,race_number,feed_time_utc,retrieved_at_utc,asset_id,skill,asset_type,value,team_id,full_name,...,fastest_lap_points,dotd_points,overtake_points,q3_finishes_points,top10_race_position_points,top8_sprint_position_points,total_position_points,total_positions_gained_lost_points,total_dnf_dq_points,value_for_money
25,2026,12,2026-09-17 13:28:13+00:00,2026-10-06 11:13:38.865379+00:00,27,2,CONSTRUCTOR,31.0,27,McLaren,...,30.0,0.0,149.0,166.0,253.0,53.0,472.0,16.0,-120.0,1.71


This disproves the theory that there is a connection between the number in the isdrstaken column and the playerpostion asset column. 

In [109]:
from pathlib import Path
import json
import pandas as pd

records = []

for file_path in Path("../../data/raw/detailed_teams").glob("*.json"):
    with open(file_path) as f:
        raw = json.load(f)

    team = raw["Data"]["Value"]["userTeam"][0]

    records.append({
        "race_number": raw["Data"]["Value"]["mdid"],
        "team_no": team["teamno"],
        "team_name": team["teamname"],
        "teamval": team.get("teamval"),
        "teambal": team.get("teambal"),
        "maxteambal": team.get("maxteambal"),
    })

values_df = pd.DataFrame(records)

values_df["teamval_plus_teambal"] = (
    values_df["teamval"] + values_df["teambal"]
)

values_df.sort_values(
    ["race_number", "team_name"]
).head(30)

,race_number,team_no,team_name,teamval,teambal,maxteambal,teamval_plus_teambal
91,1,2,BBB-TEAM,99.5,0.5,100.0,100.0
139,1,1,BoxBoxBrigade,99.9,0.1,100.0,100.0
33,1,1,Ella%20Racing%20Team,93.7,6.3,100.0,100.0
216,1,1,English%20Racing,99.8,0.2,100.0,100.0
26,1,1,Gogo%40interlagos,96.0,4.0,100.0,100.0
87,1,1,Gor%C4%85co%20Polecam,91.9,8.1,100.0,100.0
97,1,1,In%20Search%20of%20Lost%20Sainz,97.9,2.1,100.0,100.0
67,1,1,Leave%20Briatore%20alone!!,97.5,2.5,100.0,100.0
83,1,2,Minus%20Twenty%20Racing,99.5,0.5,100.0,100.0
162,1,1,No%20Michael%2C%20no!%20That's...,99.9,0.1,100.0,100.0


In [110]:
check_df = values_df.dropna(
    subset=["teamval", "teambal", "maxteambal"]
).copy()

check_df["difference"] = (
    check_df["maxteambal"]
    - check_df["teamval_plus_teambal"]
).round(2)

check_df["difference"].value_counts()

difference
0.0    128
Name: count, dtype: int64

In [111]:
values_df[
    ["teamval", "teambal", "maxteambal"]
].isna().sum()

teamval       97
teambal        0
maxteambal    97
dtype: int64

In [112]:
values_df.groupby("race_number").agg(
    teams=("team_no", "size"),
    teamval_available=("teamval", "count"),
    maxteambal_available=("maxteambal", "count"),
)

,teams,teamval_available,maxteambal_available
race_number,,,
1,15,15,15
2,15,12,12
3,15,9,9
4,15,9,9
5,15,9,9
6,15,9,9
7,15,10,10
8,15,8,8
9,15,6,6


## Conclusion / transformation plan

Existing processed CSVs:
- `asset_snapshot` = asset × race
- `league_standings_snapshot` = team × race
- `team_asset_snapshot` = team × race × asset

Use the detailed-team JSON to:

- **Enrich `league_standings_snapshot`** with race-specific budget/transfer data (`teambal`, `usersubs`, `usersubsleft`) and chip/captain information where useful.
- **Enrich `team_asset_snapshot`** with `iscaptain`, `ismgcaptain`, `playerpostion` and `isfinal`.
- **Create `team_chip_usage`** for the 6 possible chips and the race each was used.
- Keep season/current fields such as `ovpoints` and `ovrank` separate from race history (only create a season table if useful).

`teamval` is not reliable for historical team value because it may not update when a team is unchanged. Instead calculate:

**team value = sum of the 7 assets' values for that race + `teambal`**

using `team_asset_snapshot` + `asset_snapshot`.

Also remove `user_name` from processed data and use `team_name` instead.